### PR TITLE
Tgmonkey

### PR DESCRIPTION
**Pull request by Trugath**
Robust Monkeys. Tested and in-use on BPL server. Only really a problem in hordes.

They steal things, aggro, disarm and beat you to crit, then stuff your body in disposals.
Keeping the changelog sparse so it's more Fun

🆑 Monkey Management
add: Added Monkey Behaviour
/🆑

Monkeys that step and make 4 random emotes are boring. Monkeys should be unpredictable, inquisitive and vindictive little bastards.

### CODE REVIEW

**Comment by Lakart** [comment]
Genetics, Viro, and Xeno can be pretty stagnant. Having to beat a monkey to death every now and then adds spice to life.

And the kitchen is used to things that fight back when you try to make them into meat, or just because.

**Comment by Arianya** [comment]
Does this mean I can't buckle a monkey to be my blood doll in Xeno anymore?

**Comment by Arianya** [comment]
Also, looking forward to antag xenobiologist releasing a horde of monkeys

**Comment by Trugath** [comment]
The resist out of buckle doesn't work based on my testing. they resist but the code doesn't pop them out the chair. decided that is a WIP. Didn't want to stomp over too much code. They do succesfully escape cuffs in about 90 seconds though.

**Comment by PKPenguin321** [comment]
oh boy oh boy oh boy!

File: code/modules/mob/living/carbon/monkey/combat.dm [hunk 1]
```diff
@@ -0,0 +1,436 @@
...
+		if(!istype(C) || !C || in_range(src, target))
+			return
+		C.visible_message("<span class='warning'>[pick( \
+						  "[C] dives out of [src]'s way!", \
+						  "[C] stumbles over [src]!", \
+						  "[C] jumps out of [src]'s path!", \
+						  "[C] trips over [src] and falls!", \
+						  "[C] topples over [src]!", \
+						  "[C] leaps out of [src]'s way!")]</span>")
+		C.Weaken(2)
```
Comment (Shadowlight213) [inline code/modules/mob/living/carbon/monkey/combat.dm]: \>aggroed monkeys have the beepsky chasing stun.
Why?

**Comment by optimumtact** [comment]
@Cheridan

File: code/modules/mob/living/carbon/monkey/combat.dm (hunk 1 above)
Comment (Trugath) [inline code/modules/mob/living/carbon/monkey/combat.dm]: It was hilarious. A horde of monkies will trip and dive over each other to get at you.

Comment (Trugath) [inline code/modules/mob/living/carbon/monkey/combat.dm]: Feel free to remove, it wasn't in the original design. It just 'worked' nicely

Comment (Trugath) [inline code/modules/mob/living/carbon/monkey/combat.dm]: I did add the check to stop them causing the stun when they are downed fully. So the window of Fun is smallish.

**Comment by RemieRichards** [comment]
>then stuff your body in disposals

Absolutely not.

**Comment by Cheridan** [comment]
@optimumtact Very yes

**Comment by Trugath** [comment]
They will only try to disposals you if and only if: You attacked them directly or aggroed them in a way to be added to the Enemies list. You did so enough that them beating you didnt remove you from that list through the de-aggro mechanic and you are within sight of a disposals.
They do not press the handle.
This gives you time to just get out if you arn't in crit. If you are? there should be a cargo tech on duty to save you.

**Comment by RemieRichards** [comment]
>They do not press the handle.

This is not a point in the favour of the mechanic, you do realise this right? being dumped in a disposal without it going of is essentially removing the person from the round forever, nobody checks disposal bins, you're just gone if it doesn't flush, there's an auto flush, but it's 100% obvious people will start turning that off when they know that they can ruin things for other people via monkeys.

**Comment by Trugath** [comment]
Turn your suit sensors to Max. If it worries you.

**Comment by optimumtact** [comment]
Don't disposals autoflush if things are in them?

**Comment by Trugath** [comment]
The autoflush means them stuffing and not pressing the handle gives someone not in crit time to escape, someone in crit will be auto-flushed. RemieRcihards is worried about peaple disabling autoflush.

**Comment by optimumtact** [comment]
? People already can crit other people and stuff them in disposals and that may or may not have autoflush turned off.

Seems like a silly point of order to complain about imo.

**Comment by KorPhaeron** [comment]
This is great.

Hope it doesn't ruin the departments that use them though

**Comment by Trugath** [comment]
Only thing i would really suggest is for Virology: Putting a mask on a monkey stops it biting. 
Stops unwanted infections.

**Comment by PKPenguin321** [comment]
maybe make domestic monkeys which are like what we already have to use for stuff like pun pun

**Comment by Lakart** [comment]
They only hit you if you hit them, even a domesticated animal wont let you kick it around.

**Comment by Shapsy** [comment]
I really doubt people would go around turning off autoflush on disposals just in case a monkey tried stuffing someone into one

Perhaps add a lobotomy surgery for monkies that removes all the behaviors this adds and makes them docile and unable to fight back, and make Pun Pun lobotimized by default? That'd probably be a lot of snowflake code though

**Comment by AdamElTablawy** [comment]
so like goon monkeys then

fair warning-on goon, monkeys are extremely robust and routinely crit people who fuck with them.

similar things may happen here so make sure departments have straightjackets or something to avoid monkey RAGE destroying genetics

**Comment by Lakart** [comment]
Not nearly as robust as goon monkeys, you could take 3 of them bare handed pretty easily.

**Comment by AdamElTablawy** [comment]
lame

you haven't lived until you accidentally grab a monkey wrong on goon and 5 of them crash through the window suplex you onto a table and cut out your heart with a saw and scapel

**Comment by Lakart** [comment]
5 of them will probably kill you.

**Comment by MimicFaux** [comment]
I may have to give goon another chance. That sounds amazing.

**Comment by MMMiracles** [comment]
saying no to monkey rage is saying no to fun

**Comment by Armhulen** [comment]
@MimicFaux There's a special role called test subject that makes you a monkey assistant roundstart with a custom name

and monkeys love you, you can disposal dive into genetics and start VIVA LA MONKEY

File: code/modules/mob/living/carbon/human/human_defense.dm [hunk 2]
```diff
@@ -213,6 +213,18 @@
...
+		if(get_active_held_item() && drop_item())
+			playsound(loc, 'sound/weapons/slash.ogg', 25, 1, -1)
+			visible_message("<span class='danger'>[M] disarmed [src]!</span>", \
+					"<span class='userdanger'>[M] disarmed [src]!</span>")
+		else
+			playsound(loc, 'sound/weapons/pierce.ogg', 25, 1, -1)
+			Weaken(5)
+			add_logs(M, src, "tackled")
+			visible_message("<span class='danger'>[M] has tackled down [src]!</span>", \
+				"<span class='userdanger'>[M] has tackled down [src]!</span>")
```
Comment (RandomMarine) [inline code/modules/mob/living/carbon/human/human_defense.dm]: We're going to see lings going monkey just for guaranteed tackles.

Comment (Trugath) [inline code/modules/mob/living/carbon/human/human_defense.dm]: Left the disarm, the stun is less reliable for controlled monkies now.

**Comment by FantasticFwoosh** [comment]
@Armhulen now all sentient monkeys need slime-esque follows & rallies (active ability for infected monkeys?) as well as a attack so you can gather up monkeys and send them en-masse at people kicking & biting.

Really actually you could swap the 'test subject' out with anything, id personally use a Gorilla for a bit of muscle and variety (and literally nobody is going to mess with a gorilla unlike a chef chasing the monkeys) but playing a monkey assistant sounds great.

**Comment by jughu** [comment]
PM TO ADMINS:HELP A MONKEY JUST KILLED ME AND STUFFED MY BODY IN A DISPOSAL IS HE TRAITOR?
ADMIN PM TO PLAYER:no thats just how monkeys always act from now on
PM TO ADMINS:FUCKING CODERS!!!

**Comment by jughu** [comment]
I just gonnaa open 15 monkeycubes in armoury i dare anyone to attack it

**Comment by Trugath** [comment]
Aslong as the players don't attack any of them, they will more or less ignore the players. They will snatch things out of hands though.  They only aggro due to action.

**Comment by jughu** [comment]
I tottaly dont see this used against nuke ops
MONKEYTIDE!!! 
and they shoot shoot shoot shoot them untill they run out of bullets

**Comment by Trugath** [comment]
A horde of monkeys are a serious threat if you hit them accidentally.

**Comment by jughu** [comment]
Do they aggro on everything or only the attacker?

**Comment by Chocobro** [comment]
:+1:

**Comment by Trugath** [comment]
Only the attacker.

**Comment by jughu** [comment]
That sucks tbh ENRAGED MONKEYS should be hostile to everything for a while

**Comment by Trugath** [comment]
I didn't want people to monkeycube, smack and run off. I want spice and flavourful monkeys. Not hate driven murder machines.  It would be fine if they didn't swarm. With the swarm, I worry any generalised murder mechanic could be abused.

File: code/modules/mob/living/carbon/monkey/combat.dm [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+		return
+	..()
+
+/mob/living/carbon/monkey/proc/take_to_slot(obj/item/G)
+	var/list/slots = list ("left hand" = slot_hands,"right hand" = slot_hands)
```
Comment (AnturK) [inline code/modules/mob/living/carbon/monkey/combat.dm]: What's the point of these helpers, if they try same thing twice.

File: code/modules/mob/living/carbon/monkey/combat.dm [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+	G.loc = src
+	if(G.force && G.force > best_force)
+		best_force = G.force
+	equip_in_one_of_slots(G, slots)
+	// update_hands = 1
+
+/mob/living/carbon/monkey/proc/equip_in_one_of_slots(obj/item/I, list/slots, qdel_on_fail = 0)
```
Comment (AnturK) [inline code/modules/mob/living/carbon/monkey/combat.dm]: I'm pretty sure we have this on humans already, just move it to /carbon

**Review by AnturK — COMMENTED** [review]
Just quick look, will read over it later.

File: code/modules/mob/living/carbon/monkey/combat.dm [hunk 5]
```diff
@@ -0,0 +1,436 @@
...
+			if(!resisting)
+				walk_to(src,0)
+
+			return resisting
+
+		if(MONKEY_HUNT)		// hunting for attacker
+			if(health < 50)
+				mode = MONKEY_FLEE
+				spawn(1)
+					handle_combat()
```
Comment (coiax) [inline code/modules/mob/living/carbon/monkey/combat.dm]: SPAWNS ARE EVIL, USE addtimer

File: code/modules/mob/living/carbon/monkey/combat.dm [hunk 6]
```diff
@@ -0,0 +1,436 @@
...
+/mob/living/carbon/monkey/proc/back_to_idle()
+	mode = MONKEY_IDLE
+	target = null
+	a_intent = INTENT_HELP
+	frustration = 0
+
+// handle de-aggro
+/mob/living/carbon/monkey/proc/attacked(mob/living/carbon/H)
+	if(prob(25))
+		enemies[H] --
```
Comment (Trugath) [inline code/modules/mob/living/carbon/monkey/combat.dm]: Incrementing uninitilised member of a list sets it to 1. I'm using the list as a Map. Mapping from Mob to Aggression. If a mob is in the map it is an enemy. The number is used in the de-aggro mechanic.

Comment (Trugath) [inline code/modules/mob/living/carbon/monkey/combat.dm]: In the wierd case of the mob not being in enemies but attacked by the monkey, it gets added to enemies at -1 then immedietly removed. I could add a check... Okay i'll add the check

**Comment by Anonmare** [comment]
I'm sure someone could add a rage chemical later on. After that, just pass your monkey a dual e-sword, enrage them and run the hell away.

**Comment by Militaires** [comment]
@mimicfaux
come over
i suggest LLJK 1 if you like RP, 2 if otherwise, goon has some pretty fun stuff.

**Comment by Trugath** [comment]
They snatch from your hands. They don't take from the pocket slots.

**Comment by Saegrimr** [comment]
dicks out

File: code/modules/mob/living/carbon/monkey/combat.dm [hunk 7]
```diff
@@ -0,0 +1,440 @@
...
+
+/mob/living/carbon/monkey/hitby(atom/movable/AM, skipcatch = 0, hitpush = 1, blocked = 0)
+	if(istype(AM, /obj/item))
+		var/obj/item/I = AM
+		if(I.throwforce < src.health && I.thrownby && ishuman(I.thrownby))
+			var/mob/living/carbon/human/H = I.thrownby
+			retaliate(H)
+	..()
+
+/mob/living/carbon/monkey/Crossed(atom/movable/AM)
```
Comment (Trugath) [inline code/modules/mob/living/carbon/monkey/combat.dm]: Struggled with this, Was copied originally but the control flow is different between the two mobs. Best I could think of without risking breaking something was pushing the emote and weaken down into a proc in living. I called the proc knockDown but i appreciate suggestions on improving this.

**Comment by Trugath** [comment]
There is a lot of code shared between this, secbot and ed209. The code isn't easily de-duplicated because it relies on state only in these mobs. Additionally the goals and concepts of the patch (AI behaivour) are in other places in the code as well.

I guess the one long term solution is to rafactor everything into snpc/interactive and make that more general for living things as opposed to just humans?
The another is to do something with brains. Put the behaivour in them and let people switch brains around. This would let someone code the lobotomy surgery that was suggested earlier.

There is a world of options, but really I just want less boring lifeless monkeys.

**Comment by jughu** [comment]
The people who downvoted this are unrobust cucks

**Comment by AdamElTablawy** [comment]
@Trugath 
tbh i miss the days of giving monkeys an esword, punching them and RUNNING and watching the monkey murder anything that came near it on goon, so them being aggro on everything would be p funny

@MimicFaux 

you haven't lived until you see a wizard get his shit kicked in by four monkeys with knives

no seriously there was a wizard one round and he accidentally punched a monkey and like five of them ran out of genetics suplexed him to the floor and started stabbing him to death

**Comment by Trugath** [comment]
@AdamElTablawy If this get merged and people don't revolt, we can think about possible changes.

**Comment by AnturK** [comment]
How WIP is this ? I suppose testmerge would be in order.

**Comment by Trugath** [comment]
Apart from any reported bugs, balance requests and suggested tweaks? I'm done for now.

**Comment by AnturK** [comment]
Judging from testmerge, works fine. 
Could probably use a tune up in strength and some aggro on special effects (flashes and such) but that's secondary.

File: code/modules/mob/living/carbon/monkey/combat.dm [hunk 8]
```diff
@@ -0,0 +1,433 @@
+
+
+/mob/living/carbon/monkey
+	var/aggressive=0
```
Comment (AnturK) [inline code/modules/mob/living/carbon/monkey/combat.dm]: This variable is not used ?

**Comment by dannno** [comment]
Yeah honestly they aren't deadly enough
I wanted to punch a genetics monkey and be literally torn apart
also weren't they supposed to be friendly if you are also a monkey? because they fucked me up regardless

**Comment by MrStonedOne** [comment]
I'd prefer they came in too weak, then got buffed later on after some play time, then the reverse.

**Comment by Iamgoofball** [comment]
What if they used deep learning

File: code/modules/mob/living/carbon/monkey/combat.dm [hunk 9]
```diff
@@ -0,0 +1,433 @@
...
+		if(W.force >= best_force)
+			put_in_hands(W)
+			best_force = W.force
+			return TRUE
+
+	// CLOTHING
+	else if(istype(I,/obj/item/clothing))
+		var/obj/item/clothing/C = I
+		monkeyDrop(C)
+		spawn(5)
```
Comment (optimumtact) [inline code/modules/mob/living/carbon/monkey/combat.dm]: why do you have spawns here

File: code/modules/mob/living/carbon/monkey/combat.dm [hunk 10]
```diff
@@ -0,0 +1,433 @@
...
+				pickupTarget = null
+				pickupTimer = 0
+
+			// in someones hand
+			else if(ismob(pickupTarget.loc))
+				var/mob/M = pickupTarget.loc
+				if(!pickpocketing)
+					pickpocketing = TRUE
+					M.visible_message("[src] starts trying to take [pickupTarget] from [M]", "[src] tries to take [pickupTarget]!")
+					spawn(5)
```
Comment (optimumtact) [inline code/modules/mob/living/carbon/monkey/combat.dm]: another spawn?

File: code/modules/mob/living/carbon/monkey/combat.dm [hunk 11]
```diff
@@ -0,0 +1,433 @@
...
+
+	// nuh uh you don't pull me!
+	if(pulledby && (mode != MONKEY_IDLE || prob(5)))
+		if(Adjacent(pulledby))
+			a_intent = INTENT_DISARM
+			pulledby.attack_paw(src)
+			retaliate(pulledby)
+			return TRUE
+
+	// nrrr, RARRRGH
```
Comment (optimumtact) [inline code/modules/mob/living/carbon/monkey/combat.dm]: please dont' leave in commented out code

File: code/modules/mob/living/carbon/monkey/combat.dm [hunk 12]
```diff
@@ -0,0 +1,433 @@
...
+		if(!istype(M) || !M)
+			return
+		knockOver(M)
+		return
+	..()
+
+/mob/living/carbon/monkey/proc/monkeyDrop(var/obj/item/A)
+	if(A)
+		unEquip(A)
+		A.loc = get_turf(src) // drop item works inconsistently
```
Comment (optimumtact) [inline code/modules/mob/living/carbon/monkey/combat.dm]: This should never happen, and even if it does you should use A.forceMove(get_turf(src))

Comment (optimumtact) [inline code/modules/mob/living/carbon/monkey/combat.dm]: Also some items can get deleted on drop so you should check !qdeleted(A) after

File: code/modules/mob/living/carbon/monkey/combat.dm (hunk 9 above)
Comment (Trugath) [inline code/modules/mob/living/carbon/monkey/combat.dm]: These spawns are a non-blocking delay. I can look into replacing them with something else.

File: code/modules/mob/living/carbon/monkey/combat.dm (hunk 10 above)
Comment (Trugath) [inline code/modules/mob/living/carbon/monkey/combat.dm]: do_mob is a blocking call, the spawn removes the blocking call from life.

**Comment by optimumtact** [comment]
This has been sitting in review hell because personally I'm lukewarm on the pr so I don't want to review and merge it myself, and all the other maintainers are basically lazy as shit.

Sorry 😢

**Comment by RemieRichards** [comment]
I'm not lazy, I just hate the concept, and the code of it I've seen is just copypasted SNPC code.
Not worth merging tbh.

**Comment by Trugath** [comment]
It's okay. I never expected it to get merged. Someone bugged me to submit it it as a PR. 

My original plan was to isolate the new behaviour in minimal code changes to the existing code-base. Hence the 'style' of code. I got the desired behaviour working using accepted code from the codebase as templates. The main part of the code is the state machine, everything around it is supporting code and is prime target for refactoring.

Rewriting it to obscure the fact I did template it from interactive and secbot would have just gone against any refactoring effort later.

I tried dropping into the irc channel to discuss the PR and it's code a few times. I didn't get any response, which is fine, i understand it might have been bad timing. I was sadly asleep (I'm on GMT) both times I know it was test merged.

This was a tentative 'Trying not to step on maintainers toes' PR. I can be more aggressive with the code if required.

**Comment by RemieRichards** [comment]
I actually wrote a review just now but misclicked and closed chrome (fuuuuu), the gist of it is this:
* ```spawn()``` -> ```addtimer()``` (look into this proc)
* Defines/Consts for magic values
* ```locate()```s are bad so try to store them in vars instead of redoing them (there's atleast 2-3 cases in here I've seen)

**Comment by RemieRichards** [comment]
Much nicer.

File: code/modules/mob/living/carbon/monkey/combat.dm (hunk 8 above)
Comment (AnturK) [inline code/modules/mob/living/carbon/monkey/combat.dm]: Still not used. Mostly want this implemented so admin created monkey hordes are dangerous. But removal also works.

**Comment by AnturK** [comment]
@Trugath Is it ready ? The major issues are gone, but don't know if you're done with the changes.

**Comment by Trugath** [comment]
It's done. I'll hands off till someone complains about something.

**Comment by Cyberboss** [comment]
Inb4 geneticists/viros valid all monkeys

**Pull request merged**